compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2284,21 +2287,77 @@ public void visit(BLangRecordLiteral recordLiteral) {
+                boolean isSpreadExprRecordTypeSealed = spreadExprRecordType.sealed;
+                if (!isSpreadExprRecordTypeSealed) {
+                    // More than one inclusive-type-descriptors with spread-fields are not allowed.
+                    if (inclusiveTypeSpreadField != null) {
+                        this.dlog.error(spreadOpExpr.pos, DiagnosticCode.MULTIPLE_INCLUSIVE_TYPES);
+                    } else {
+                        inclusiveTypeSpreadField = spreadOpField;
+                    }
+                }
+
+                LinkedHashMap<String, BField> fieldsInRecordType = spreadExprRecordType.fields;
+                for (Object fieldName : names) {
+                    if (fieldsInRecordType.containsKey(fieldName)) {
+                        if (fieldsInRecordType.get(fieldName).type.tag != TypeTags.NEVER) {
+                            this.dlog.error(spreadOpExpr.pos, DiagnosticCode.DUPLICATE_KEY_IN_RECORD_LITERAL_SPREAD_OP,
+                                            recordLiteral.type.getKind().typeName(),
+                                            fieldName,
+                                            spreadOpField);
+                        }
+                    } else {
+                        if (!isSpreadExprRecordTypeSealed) {
+                            this.dlog.error(spreadOpExpr.pos, DiagnosticCode.ALREADY_SPECIFIED_KEYS_IN_SPREAD_FIELD,
+                                            spreadOpExpr);
+                            break;
+                        }
                     }
+                }
 
+                for (BField bField : fieldsInRecordType.values()) {
                     String name = bField.name.value;
-                    if (names.contains(name)) {
-                        this.dlog.error(spreadOpExpr.pos, DiagnosticCode.DUPLICATE_KEY_IN_RECORD_LITERAL_SPREAD_OP,
-                                        recordLiteral.expectedType.getKind().typeName(), name, spreadOpField);
+                    if (!names.contains(name)) {
+                        if (bField.type.tag == TypeTags.NEVER) {
+                            neverTypedKeys.add(name);
+                        } else {
+                            if (!neverTypedKeys.remove(name) &&
+                                    inclusiveTypeSpreadField != null && isSpreadExprRecordTypeSealed) {
+                                this.dlog.error(spreadOpExpr.pos,
+                                                DiagnosticCode.DUPLICATE_KEY_IN_RECORD_LITERAL_SPREAD_OP,
+                                                recordLiteral.expectedType.getKind().typeName(),
+                                                name,
+                                                spreadOpField);
+                            }
+                            names.add(name);
+                        }
                     }
-                    names.add(name);
                 }
```

Comment:
Shall we reorganize this to reduce indentation?

```ballerina
                for (BField bField : fieldsInRecordType.values()) {
                    String name = bField.name.value;
                    if (names.contains(name)) {
                        continue;
                    }

                    if (bField.type.tag == TypeTags.NEVER) {
                        neverTypedKeys.add(name);
                        continue;
                    }

                    if (!neverTypedKeys.remove(name) &&
                            inclusiveTypeSpreadField != null && isSpreadExprRecordTypeSealed) {
                        this.dlog.error(spreadOpExpr.pos,
                                        DiagnosticCode.DUPLICATE_KEY_IN_RECORD_LITERAL_SPREAD_OP,
                                        recordLiteral.expectedType.getKind().typeName(),
                                        name,
                                        spreadOpField);
                    }
                    names.add(name);
                }
```

Reply:
Thanks